File: Solution/【025】K个一组翻转链表/K个一组翻转链表.cpp
/*
(1)

*/




class Solution {
public:
    ListNode* reverseKGroup(ListNode* head, int k) {
        if(head == nullptr) return nullptr;
        if(k <= 0)  return nullptr;
        if(k == 1)  return head;
        ListNode* new_head = new ListNode(0);
        new_head->next = head;
        ListNode* temp = new_head;
        while(temp->next != nullptr){
            temp = reverseSingleK(temp, k);
        }
        return new_head->next;
    }
    ListNode* reverseSingleK(ListNode* before, int k){   // 每次翻转 before 之后的 k 个
        ListNode* pre = before->next;     // pre 是 K 个中的第一个
        ListNode* cur = pre->next;          // cur 是 K 个中的第二个
        int flag = 0;
        for(int i = 1 ; i < k ; i ++){    // pre 是 K 个中的第一个，所以从 1 开始
            if(cur == nullptr){
                flag = i;
                break;
            }
            ListNode* temp = cur->next;     //  temp 用来暂存 cur->next，下面四行都是 头插法
            cur->next = pre;            
            pre = cur;
            cur = temp;
        }
        if(flag != 0){            //  不足 k 的再进行一次翻转
            ListNode* result = pre;
            cur = pre->next;
            for(int i = 2; i < flag; i ++){
                ListNode* temp = cur->next;
                cur->next = pre;
                pre = cur;
                cur = temp;
            }
            result->next = nullptr;
            return result;
        }
        else{                   // 够 k 个就不进行翻转
            ListNode* result = before->next;
            before->next = pre;
            result->next = cur;
            return result;
        }
    }
};

File: Solution/【025】K个一组翻转链表/K个一组翻转链表.java
/*
尾插法，用 tail 记录要翻转的部分的最后一个元素。
在遍历翻转部分元素的时候，每次把当前元素移到 tail 后面
关键是用 count 记录 tail 能走的步数，如果 步数 < k 则不进行翻转了
*/




class Solution {
    public ListNode reverseKGroup(ListNode head, int k) {
        ListNode new_head = new ListNode(0);
        new_head.next = head;
        ListNode pre = new_head;
        ListNode tail = new_head;
        while (true) {
            int count = 0;
            while (tail != null && count != k) {   // 为了找到 tail
                count++;
                tail = tail.next;
            }
            if (tail == null)   break;    // 判断 tail 是否为 空
            ListNode head1 = pre.next;
            while (pre.next != tail) {      // 尾插法，pre 在进入循环前的初始值为 需要翻转部分 的前一个
                ListNode cur = pre.next;
                pre.next = cur.next;
                cur.next = tail.next;
                tail.next = cur;
            }
            pre = head1;     // 更新 pre 和 tail
            tail = head1;
        }
        return new_head.next;
    }
}